### PR TITLE
feat: color source/target split

### PR DESCRIPTION
**Pull request by DatCaptainHorse**
## Description

Based/builds off PR #37 (required to be merged before this PR).

Per issue #36 for this one as well, but for the color handling as this includes API breaking changes, it makes more sense as a separate PR.

Basically the current API feels bit detached and allows for results that don't really make sense, such as `ColorConverterConfig::full_range` being set to `true` without calling `ColorDescription.with_full_range(true)` in combination, results in full-range output but the metadata falsely claims it as TV-range.. causing color issues when decoding.

So instead of doing this:
```rs
let mut config = ColorConverterConfig::new(1920, 1080, InputFormat::BGRx, OutputFormat::P010);
config.color_space = ColorSpace::SrgbToBt2020Pq;

let mut converter = ColorConverter::new(context.clone(), config)?;
// converter.convert(input_image, output_buffer)?;
...
```

the color config is set as so:
```rs
let config = ColorConverterConfig::new(1920, 1080, InputFormat::BGRx, OutputFormat::P010)
    .with_source(SourceColor::srgb())
    .with_target(TargetColor::Bt2020Pq)
    .with_range(ColorRange::Full);
    
let mut converter = ColorConverter::new(context.clone(), config)?;
let mut encoder = Encoder::new(context, encode_config)?;
// converter.convert(input_image, layout, encoder.input_image())?;
```

`Source -> Target` logic, bit clearer I'd say?

WIP/RFC @hgaiser - feel free to suggest changes or reject the API changes outright.

### CODE REVIEW

**Comment by DatCaptainHorse** [comment]
@hgaiser any objections to this idea? Or changes wanted? Just poking since not sure if you've seen this yet :slightly_smiling_face:

**Comment by hgaiser** [comment]
I haven't had time to look into this yet, sorry :)

File: src/converter/mod.rs [hunk 1]
```diff
@@ -265,29 +494,54 @@ impl ColorConverter {
...
     ///
-    /// This takes effect on the next `convert()` call without recreating the pipeline,
-    /// since the full-range flag is passed via push constants.
-    pub fn set_full_range(&mut self, full_range: bool) {
-        self.config.full_range = full_range;
+    /// Returns an error, and leaves the converter untouched, if the configured
+    /// source cannot reach the new target. Callers that pass
+    /// [`ColorConverterConfig::color_description`] to the encoder need to
+    /// re-declare it after this: the target is half of what that derives from,
+    /// and an encoder's session parameters were fixed when it was built.
```
Comment (hgaiser) [inline src/converter/mod.rs]: Is this true given https://github.com/hgaiser/pixelforge/blob/main/src/encoder/codec.rs#L363 ?

File: src/converter/mod.rs
Comment (hgaiser) [inline src/converter/mod.rs]: References to old ColorSpace

File: tests/color_conversion.rs
Comment (hgaiser) [inline tests/color_conversion.rs:1]: Should this file be removed / merged into tests?

File: examples/convert_quantization.rs
Comment (hgaiser) [inline examples/convert_quantization.rs:1]: Same as this one.

File: src/converter/mod.rs [hunk 2]
```diff
@@ -8,27 +8,161 @@
...
+    /// surface is.
+    Bt2020Pq,
+}
+
+impl Default for SourceColor {
+    /// [`SourceColor::srgb`].
+    fn default() -> Self {
+        Self::srgb()
+    }
+}
```
Comment (hgaiser) [inline src/converter/mod.rs]: I'm leaning to removing the default and changing where we relied on it. It doesn't make sense to assume `sRGB` anywhere we use `SourceColor` in my opinion. Do you agree?

File: src/lib.rs [hunk 3]
```diff
@@ -179,31 +179,66 @@
...
-//!
-//! | Color Space | Description |
-//! |-------------|-------------|
-//! | `Bt709` | Standard SDR (BT.709 coefficients) |
-//! | `Bt2020` | HDR passthrough (BT.2020 coefficients, PQ-encoded input) |
-//! | `SrgbToBt2020Pq` | SDR-in-HDR (sRGB → linear → BT.2020 gamut → PQ OETF) |
-//! | `Bt709LinearToBt2020Pq` | scRGB HDR (linear BT.709 → BT.2020 gamut → PQ OETF). `sdr_reference_white_nits` sets the interpretation of 1.0; per the scRGB spec (IEC 61966-2-2), 80 nits. |
+//! PixelForge includes a GPU compute shader for converting RGB input to YUV
+//! output. The conversion is two decisions: what the input already is, and
+//! what the encoded stream should be.
```
Comment (hgaiser) [inline src/lib.rs]:
```suggestion
//! output. The source color describes the input. The target color describes the stream.
```

File: src/lib.rs [hunk 4]
```diff
@@ -179,31 +179,66 @@
...
+//!
+//! | `TargetColor` | What the stream is |
+//! |---------------|--------------------|
+//! | `Bt709` | BT.709 primaries, transfer and matrix. SDR. |
+//! | `Bt2020Pq` | BT.2020 primaries, PQ transfer, BT.2020 NCL matrix. HDR10. |
+//!
+//! The relative sources carry the luminance that a sample value of 1.0 stands
+//! for, since that is a property of the source and not of the target: 203 nits
+//! for `Srgb` per ITU-R BT.2408, and 80 for scRGB per IEC 61966-2-2. It is read
+//! only on the way to `Bt2020Pq`, where the PQ encode needs it to be absolute.
```
Comment (hgaiser) [inline src/lib.rs]: I'm struggling to find why I think this reads so poorly, but it just feels very unnecessarily complicated. Maybe just AI things 🤷

**Review by hgaiser — COMMENTED** [review]
Alright took me some time to wrap my head around the situation and what you're proposing but I think I got most of it. To summarize in my own words: you identified that `ColorSpace` was used in different contexts and carried different meanings, so you opted to split it into a `SourceColor` and `TargetColor` enum, correct? I think that's a very valid point.

I like the approach, I think it makes sense. I briefly checked other libraries (libavcodec mainly). They don't seem to have a source / target split but rather enums / fields like `AVPixelFormat`, `AVColorRange`, `AVColorSpace`, `AVColorPrimaries`. What made you decide to use a split of `SourceColor` / `TargetColor`? What if we want bidirectional conversion at some point in the future? Then source/target becomes ambiguous.

File: src/converter/mod.rs (hunk 2 above)
Comment (DatCaptainHorse) [reply src/converter/mod.rs]: Agreed, will do that.

File: src/lib.rs (hunk 4 above)
Comment (DatCaptainHorse) [reply src/lib.rs]: Fair, I'll clarify it myself :+1:

**Comment by DatCaptainHorse** [comment]
> What made you decide to use a split of SourceColor / TargetColor? What if we want bidirectional conversion at some point in the future? Then source/target becomes ambiguous.

It was mostly to keep the cases with "odd" source/target combos limited and reduce edge-cases, but yeah it does add unneeded complexity to the API and limits it. I'll change things to be more unified, will need more tests for the various combinations but this is a draft PR still, will work on it :+1:

**Comment by DatCaptainHorse** [comment]
@hgaiser unified the source/target colors into `ColorSpec` enum. `ColorConverter` checks the given color params if the shader supports their conversion, returning error result explaining why not.

It's still 2 parameters in `new`, if that's too much, another option I see is unifying `InputFormat` and `OutputFormat` like the source/target colors were here and then unify them both under some kind of generic format config on top..

Lots of options, let me know what you think :slightly_smiling_face:

**Review by hgaiser — APPROVED** [review]
Looks good to me :)

**Comment by DatCaptainHorse** [comment]
@hgaiser I don't think this PR needs anything else, feel free to merge when available, I got plenty more PRs incoming :P

**Comment by hgaiser** [comment]
> @hgaiser I don't think this PR needs anything else, feel free to merge when available, I got plenty more PRs incoming :P

Oh dear (but also <3). I'll be away for about two weeks so might take a while :p
